File: rewrite/examples/yield.go
package main

import "runtime"
import "sync/atomic"

const num_tasks = 100 * 1000
const num_yields = 200

func main() {
    var counter uintptr = 0
    done := make(chan struct{}, 1)

    for i := 0; i < num_tasks; i++ {
        go func(){
            for j := 0; j < num_yields; j++ {
                runtime.Gosched()
            }
            
            if atomic.AddUintptr(&counter, 1) == num_tasks {
                done <- struct{}{}
            }
        }()
    }

    <- done
    if counter != num_tasks {
        panic("Bad counter")
    }
}

File: rewrite/examples/yield.rs
use std::sync::{Arc, atomic::{AtomicUsize, Ordering}};

const NUM_TASKS: usize = 100 * 1000;
const NUM_YIELDS: usize = 200;

#[tokio::main]
async fn main() {
    let counter = Arc::new(AtomicUsize::new(0));
    let (tx, mut rx) = tokio::sync::mpsc::channel(1);

    for _ in 0..NUM_TASKS {
        let mut tx = tx.clone();
        let counter = counter.clone();

        tokio::spawn(async move {
            for _ in 0..NUM_YIELDS {
                tokio::task::yield_now().await;
            }
            
            if counter.fetch_add(1, Ordering::Relaxed) + 1 == NUM_TASKS {
                tx.send(()).await.unwrap();
            }
        });
    }

    rx.recv().await.unwrap();
    assert_eq!(counter.load(Ordering::Relaxed), NUM_TASKS);
}